src/Components/Logs/Logs.js

```diff
@@ -26,30 +26,71 @@ import i18n from "../../i18n/i18n";
 import { useStyles } from "./styles";
 import "./Logs.css";
 
+function transformLog(log, _index, _data, ts_multiplier = 1000) {
+  const timestamp = ts_multiplier * log.time;
+  const date = new Date(timestamp);
+  return ({
+    ...log,
+    timestamp,
+    time: date.toLocaleTimeString(),
+    date: date.toLocaleDateString(),
+    key: log.message + timestamp,
+  });
+}
+
 function logsDedupe(oldLogs, data) {
-  if (!oldLogs.length)
-    return data.length;
+  if (!data.length)
+    return oldLogs;
 
-  const oldDate = oldLogs[0].timestamp;
-  const oldKey = oldLogs[0].key;
-  let j;
+  const oldDate = data[data.length - 1].time * 1000;
+  let i;
 
-  // starting from oldest new log, compare with newest old log.
-  // decrease j until we find a log that is not present
+  let map = {};
 
-  for (j = data.length - 1; j > -1 ; j--) {
-    const newDate = data[j].time;
-    const newKey = data[j].message + newDate;
+  // iter over old logs with last timestamp of the new logs
+  // and put in a map. Store the index.
 
-    if (newDate > oldDate || (
-      newDate === oldDate && newKey !== oldKey
-    ))
-      break;
+  for (
+    i = 0;
+    i < oldLogs.length && oldLogs[i].timestamp === oldDate;
+    i++
+  ) {
+    const log = oldLogs[i];
+    map[log.message] = { ...log, index: i };
   }
 
-  return j;
+  // final i is the index of the oldest
+  // log of that timestamp in old logs
+
+  let k = 0, z;
+
+  // iter over new logs (oldest to latest) with last timestamp,
+  // check if present in last map
+  //  - if present, check the index and set k to it.
+  //  - if not, add to map, set index to k - i. decrease k;
+  for (
+    k = 0, z = data.length;
+    z > 0 && data[z - 1].time === oldDate;
+    z--
+  ) {
+    const log = transformLog(data[z - 1]);
+    const exist = map[log.message];
+    if (!exist) {
+      map[log.message] = { ...log, index: k - i };
+      k--;
+    }
+  }
+
+  // cut new logs up to z, cat with the deduped ones
+  // and the old logs up to i
+  return [].concat(
+    data.slice(0, z).map(log => transformLog(log)),
+    Object.values(map).sort(({ index: a }, { index: b }) => a - b),
+    oldLogs.slice(i),
+  );
```

Comment:
do we need the empty array here?

Would this work all the same?

```suggestion
  return data.slice(0, z).map(log => transformLog(log)).concat(
    Object.values(map).sort(({ index: a }, { index: b }) => a - b),
    oldLogs.slice(i),
  );
```

Reply:
I think it would still work, yes. Should I change it?

Reply:
Yeah. Test it before shipping tho! Good job.